core/domain/user_services.py

```diff
@@ -850,19 +922,70 @@ def create_new_user(gae_id, email):
     Raises:
         Exception. If a user with the given gae_id already exists.
     """
-    user_settings = get_user_settings(gae_id, strict=False)
+    user_settings = get_user_settings_by_gae_id(gae_id, strict=False)
     if user_settings is not None:
-        raise Exception('User %s already exists.' % gae_id)
+        raise Exception(
+            'User %s already exists for gae_id %s.'
+            % (user_settings.user_id, gae_id))
 
     user_id = user_models.UserSettingsModel.get_new_id('')
     user_settings = UserSettings(
         user_id, gae_id, email, feconf.ROLE_ID_EXPLORATION_EDITOR,
         preferred_language_codes=[constants.DEFAULT_LANGUAGE_CODE])
+    _save_user_auth_details(UserAuth(user_id, gae_id))
     _save_user_settings(user_settings)
     create_user_contributions(user_id, [], [])
     return user_settings
 
 
+def _save_user_auth_details(user_auth):
+    """Commits a user auth object to the datastore.
+
+    Args:
+        user_auth: UserAuth. The user auth domain object to be saved.
+    """
+    user_auth.validate()
+
+    user_auth_dict = {
+        'gae_id': user_auth.gae_id,
+        'pin': user_auth.pin,
+        'parent_user_id': user_auth.parent_user_id,
+        'deleted': user_auth.deleted
+    }
+
+    # If user auth entry with the given user_id does not exist, create a new
+    # one.
+    user_auth_model = user_models.UserAuthModel.get_by_id(user_auth.user_id)
+    if user_auth_model is not None:
+        user_auth_model.populate(**user_auth_dict)
+        user_auth_model.put()
+    else:
+        user_auth_dict['id'] = user_auth.user_id
+        user_models.UserAuthModel(**user_auth_dict).put()
+
+
+def _get_user_auth_from_model(user_auth_model):
+    """Transform user auth storage model to domain object.
+
+    Args:
+        user_auth_model: UserAuthModel or None. The model to be converted or
```

Comment:
Why is None valid? Why not just expect that the caller ensures that the model is defined?

If 'None' is really what we want, we should define what it means for/what happens when 'None' is used.

Reply:
Actually I was trying to mimic the design of `user_services._transform_user_settings()`- [this](https://github.com/oppia/oppia/blob/develop/core/domain/user_services.py#L748). We can either do a check before sending the model or do it like the current implementation - after getting the domain object. Both the implementations styled are equally good in my opinion. It just depends where we want to put the if condition. If remove None inside this function seems like a better coding style, then I will go with that.

(Purpose was just that in the domain layer we will mostly deal with the UserAuth domain object, so instead of requiring an additional check before we convert it (whether is model is None or not), we will first convert storage object to domain object and then see. One analogous example for `UserSettings` is [this](https://github.com/oppia/oppia/blob/develop/core/domain/user_services.py#L512).)

Reply:
I think this is similar to one of my other comments (possibly in a different PR) -- if the function conceptually is dealing with an actual model, then callers should pass in that model rather than "model or None". In this case I think callers should pass in an actual model.

I suggest updating the existing UserSettings functions you are copying in order to remove this bad pattern, so that others don't copy it in the future too. (This code was written a long time ago in a time where our codebase was less structured, so we probably ought to update it.)

Also, thanks for explaining your reasoning so clearly!

Reply:
@seanlip thanks for clarifying. I will do the changes.

Reply:
Did the changes in `user_services._transform_user_settings()` as well as `_save_user_auth_details` - removed None inside these functions and changed at code at related places.